src/ray/raylet/node_manager.cc

```diff
@@ -2969,22 +2969,46 @@ void NodeManager::ProcessSubscribePlasmaReady(
   auto message = flatbuffers::GetRoot<protocol::SubscribePlasmaReady>(message_data);
   ObjectID id = from_flatbuf<ObjectID>(*message->object_id());
 
-  // Tell task dependency manager to pull the objects.
-  // NOTE(simon): This doesn't handle cancellation and edge case invovling eviciton.
-  std::vector<rpc::ObjectReference> refs = {FlatbufferToSingleObjectReference(
-      *message->object_id(), *message->owner_address())};
-  task_dependency_manager_.SubscribeWaitDependencies(associated_worker->WorkerId(), refs);
-
-  {
-    absl::MutexLock guard(&plasma_object_notification_lock_);
-    if (!async_plasma_objects_notification_.contains(id)) {
-      async_plasma_objects_notification_.emplace(
-          id, absl::flat_hash_set<std::shared_ptr<WorkerInterface>>());
-    }
+  if (task_dependency_manager_.CheckObjectLocal(id)) {
+    // Object is already local, so we directly fires the callback to tell core worker
+    // plasma object is ready.
+    rpc::PlasmaObjectReadyRequest request;
+    request.set_object_id(id.Binary());
+
+    RAY_LOG(DEBUG) << "Object " << id << " is already local, firing callback directly.";
+    associated_worker->rpc_client()->PlasmaObjectReady(
+        request, [](Status status, const rpc::PlasmaObjectReadyReply &reply) {
+          if (!status.ok()) {
+            RAY_LOG(INFO) << "Problem with telling worker that plasma object is ready"
+                          << status.ToString();
+          }
+        });
+  } else {
+    // The object is not local, so we are subscribing to pull and wait for the objects.
+    std::vector<rpc::ObjectReference> refs = {FlatbufferToSingleObjectReference(
+        *message->object_id(), *message->owner_address())};
+
+    // NOTE(simon): This call will issue a pull request to remote workers and make sure
+    // the object will be local.
+    // 1. We currently do not allow user to cancel this call. The object will be pulled
+    //    even if the `await object_ref` is cancelled.
+    // 2. We currently do not handle edge cases with object eviction where the object
+    //    is evicted locally or on remote machine.
+    task_dependency_manager_.SubscribeWaitDependencies(associated_worker->WorkerId(),
+                                                       refs);
+
+    // Add this worker to listener for the object id.
+    {
+      absl::MutexLock guard(&plasma_object_notification_lock_);
+      if (!async_plasma_objects_notification_.contains(id)) {
```

Comment:
FYI, I think you can replace all this code with just `async_plasma_objects_notification_[id].insert(associated_worker)` (C++ will dedupe shared_ptrs).